content/blog/angular-components-control-value-accessor/index.md

```diff
@@ -0,0 +1,419 @@
+
+These classes include:
+
+
+- `ng-pristine`
+- `ng-dirty`
+- `ng-untouched`
+- `ng-touched`
+
+They reflect states so that you can update the visuals in CSS to reflect them. When using `[(ngModel)]`, they won't appear, since nothing is tracking when a component is `pristine` or `dirty`. However, when using `[formControl]` or `[formControlName]`, these classes _will_ appear and act accordingly, thanks to the `registerOnChange` and `registerOnTouched` functions. As such, you're able to display custom CSS logic for when each of these values appears.
+
+# Gain Access To Form Control Errors {#form-control-errors}
+
+Something you'll notice that wasn't implemented in the `ControlValueAccessor` implementation is support for checking whether validators are applied. If you're a well-versed Angular Form-ite, you'll recall the ability to [validate forms using validators appended to `FormControl`s](https://angular.io/guide/form-validation). Although a niche situation — since most validation happens at the page level, not the component level — wouldn't it be nice to check when a form is valid or not directly from the component to which the form is attached?
+
+Well, thanks to Angular's DI system, we can do just that!
+
+However, we'll need to make a few changes to the form input [that we made before](#forwardRef). While we previously implemented a provider for form controls, we now need to manually assign the provider ourselves in the constructor:
+
+```typescript
+import {
+  Component,
+  Input,
+  ChangeDetectorRef,
+  Optional,
+  Self,
+  AfterContentInit
+} from "@angular/core";
+import { ControlValueAccessor, NgControl } from "@angular/forms";
+
+@Component({
+  selector: "app-example-input",
+  templateUrl: "./example-input.component.html",
+  styleUrls: ["./example-input.component.css"]
+})
+export class ExampleInputComponent implements ControlValueAccessor, AfterContentInit {
+  constructor(
+    @Optional() @Self() public ngControl: NgControl,
+    private _changeDetector: ChangeDetectorRef
+  ) {
+    if (ngControl != null) {
+      // Setting the value accessor directly (instead of using
+      // the providers) to avoid running into a circular import.
+      ngControl.valueAccessor = this;
+    }
+  }
+    
+  // ...
+}
+```
+
+In this code sample, we're using [the `@Self` decorator](https://angular.io/api/core/Self) to tell the dependency injection system that "this component _itself_ should have been provided a `formControl` or `formControlName`". However, we want the component to work even when `FormModule` isn't being used, so we allow the dependency injection to return `null` if nothing's passed by utilizing [the `@Optional` decorator](https://angular.io/api/core/Optional).
+
+Now that you have the `ngControl`, you can access the `formControl` by using `ngControl.control`.
+
+```typescript
+ngOnInit() {
+    const control = this.ngControl && this.ngControl.control;
```

Comment:
since this is TS, why not use the optional chain?
```suggestion
    const control = this.ngControl?.control;
```

Reply:
We _could_ use optional chaining, but it's a fairly new feature to Angular (9), so I'd like to keep it as-is